### ISSUE TITLE
Understanding the behavior of CuDNNBatchNorm

### ISSUE THREAD

**Issue by ArtanisCV** [dmlc/mxnet#3463]
Recently I met a problem that, during training, the validation accuracy is always low, although the training accuracy is high. 
There are many similar issues, like #3032, #3424, #3378. I notice that all these issues are related to batch normalization. Hence I do the following simple experiment:

    import os
    import numpy as N
    import mxnet as mx
    
    
    def build_nn():
        X = mx.symbol.Variable(name="data")
        Y = mx.symbol.Variable(name="label")
    
        net = mx.symbol.FullyConnected(data=X, num_hidden=1024)
        net = mx.symbol.LeakyReLU(data=net, act_type="leaky")
        net = mx.symbol.BatchNorm(data=net)
    
        net = mx.symbol.FullyConnected(data=net, num_hidden=2)
        return mx.symbol.SoftmaxOutput(net, Y, normalization="batch")
    
    
    def accuracy(label, pred):
        pred = pred.ravel()
        label = label[:, 1].ravel()
        return ((pred > 0.5) == label).mean()
    
    
    if __name__ == "__main__":
        batch_size = 16
        rand = N.random.RandomState(123)
    
        train_x = N.zeros((160, 10), dtype=N.float32)
        train_x[:80, :5] = rand.normal(size=(80, 5))
        train_x[80:, 5:] = rand.normal(size=(80, 5))
    
        train_y = N.zeros((160, 2), dtype=N.float32)
        train_y[:80, 0] = 1
        train_y[80:, 1] = 1
    
        train_iter = mx.io.NDArrayIter(
            data=train_x, label={"label": train_y},
            batch_size=batch_size, shuffle=True,
            last_batch_handle="pad"
        )
    
        test_x = N.zeros((16, 10), dtype=N.float32)
        test_x[:8, :5] = rand.normal(size=(8, 5))
        test_x[8:, 5:] = rand.normal(size=(8, 5))
    
        test_y = N.zeros((16, 2), dtype=N.float32)
        test_y[:8, 0] = 1
        test_y[8:, 1] = 1
    
        test_iter = mx.io.NDArrayIter(
            data=test_x, label={"label": test_y},
            batch_size=batch_size, shuffle=False,
            last_batch_handle="discard"
        )
    
        ctx = mx.gpu(0) if os.path.exists("/usr/bin/nvidia-smi") else mx.cpu()
    
        net = build_nn()
        net = mx.module.Module(symbol=net, data_names=("data",), label_names=("label",),
                               context=ctx)
        net.bind(data_shapes=train_iter.provide_data,
                 label_shapes=train_iter.provide_label)
        net.init_params(initializer=mx.initializer.Xavier(factor_type="in", magnitude=2))
        net.init_optimizer(optimizer="adam", optimizer_params=(
            ("learning_rate", 1e-4), ("wd", 0), ("rescale_grad", 1.0 / batch_size))
        )
    
        accuracy = mx.metric.CustomMetric(accuracy, name="accuracy")
    
        for epoch in xrange(20):
            train_iter.reset()
            for batch in train_iter:
                net.forward(batch, is_train=True)
                net.backward()
                net.update()
    
        test_iter.reset()
        for batch in test_iter:
            net.forward(batch, is_train=False)
            accuracy.update(batch.label, net.get_outputs())
    
        print "Validation-%s=%f" % accuracy.get()

This toy example trained a two-class softmax to classify some random-generated examples.

I run the above example in two situation:
1. ctx=mx.cpu()
2. ctx=mx.gpu(0) (with cuDNN v5)
In the first situation, everything is ok, both the training and the validation accuracy are closed to 1.
In the second situation, the training accuracy is closed to 1, but the validation accuracy is 0.5, which means the network doesn't output correctly when is_train=False.

Considering that Situation 1 used BatchNorm Op as the backend, and Situation 2 used CuDNNBatchNorm Op as the backend, is there any special thing that I should pay attention to for CuDNNBatchNorm?

**Comment by ArtanisCV** [comment]
Well, I re-run the codes on another machine, but everything is OK. Seems that there may be some other things that I didn't pay attention to.

Nevertheless, when I change the validation code to forward(is_train=True), the validation accuracy becomes consistent with the training accuracy (also works for the three issues referred above). I know that it is illegal to use batch mean / variance when testing, but seems the moving mean / variance have some strange behaviors when testing.

I will do more experiments. Sorry for disturbing you.

**Comment by George19890617** [comment]
It's all equal 1 in gnu and cpu on my machine when set is_train=False in Validation.
